File: app/src/main/java/gs/weather/engine/ThingManager.java
package gs.weather.engine;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

import javax.microedition.khronos.opengles.GL10;

public class ThingManager {
    private static final int MAX_THINGS = 64;
    private static final String TAG = "GL Engine";
    private volatile List<Thing> thingList = new ArrayList<>();

    public synchronized boolean add(Thing thing) {
        return thingList.add(thing);
    }

    public synchronized boolean clear() {
        thingList.clear();
        return false;
    }

    public synchronized boolean clearByTargetname(String name) {
        for (Iterator<Thing> it = thingList.iterator(); it.hasNext(); ) {
            Thing thing = it.next();

            if (TextUtils.equals(thing.targetName, name)) {
                it.remove();
            }
        }
        return false;
    }

    public synchronized int countByTargetname(String name) {
        int count = 0;
        for (Thing thing : thingList) {
            if (TextUtils.equals(thing.targetName, name)) {
                count++;
            }
        }
        return count;
    }

    public synchronized void render(GL10 gl10, TextureManager texMagr, MeshManager meshMagr) {
        for (Thing thing : thingList) {
            thing.renderIfVisible(gl10, texMagr, meshMagr);
        }
    }

    public synchronized void sortByY() {
        Collections.sort(thingList, new Comparator<Thing>() {
            @Override
            public int compare(Thing lhs, Thing rhs) {
                return (int) (lhs.origin.y - rhs.origin.y);
            }
        });
    }

    public synchronized void update(float timeDelta) {
        update(timeDelta, false);
    }

    public synchronized void update(float timeDelta, boolean onlyVisible) {
        for (Iterator<Thing> it = thingList.iterator(); it.hasNext(); ) {
            Thing thing = it.next();

            if (thing.isDeleted()) {
                it.remove();

            } else if (onlyVisible) {
                thing.updateIfVisible(timeDelta);

            } else {
                thing.update(timeDelta);
            }
        }
    }

}

File: app/src/main/kotlin/io/github/gmazzo/android/livewallpaper/weather/engine/ThingManager.kt
package io.github.gmazzo.android.livewallpaper.weather.engine

import android.text.TextUtils
import io.github.gmazzo.android.livewallpaper.weather.engine.things.Thing
import javax.microedition.khronos.opengles.GL10

class ThingManager {
    private val thingList: MutableList<Thing> = ArrayList()

    @Synchronized
    fun add(thing: Thing): Boolean {
        return thingList.add(thing)
    }

    @Synchronized
    fun clear(): Boolean {
        thingList.clear()
        return false
    }

    @Synchronized
    fun clearByTargetname(name: String?): Boolean {
        val it = thingList.iterator()
        while (it.hasNext()) {
            val thing = it.next()

            if (TextUtils.equals(thing.targetName, name)) {
                it.remove()
            }
        }
        return false
    }

    @Synchronized
    fun countByTargetname(name: String?): Int {
        var count = 0
        for (thing in thingList) {
            if (TextUtils.equals(thing.targetName, name)) {
                count++
            }
        }
        return count
    }

    @Synchronized
    fun render(gl10: GL10) {
        for (thing in thingList) {
            thing.renderIfVisible(gl10)
        }
    }

    @Synchronized
    fun sortByY() {
        thingList.sortBy { it.origin.y }
    }

    @Synchronized
    fun update(timeDelta: Float) {
        update(timeDelta, false)
    }

    @Synchronized
    fun update(timeDelta: Float, onlyVisible: Boolean) {
        val it = thingList.iterator()
        while (it.hasNext()) {
            val thing = it.next()

            if (thing.isDeleted) {
                it.remove()
            } else if (onlyVisible) {
                thing.updateIfVisible(timeDelta)
            } else {
                thing.update(timeDelta)
            }
        }
    }

    companion object {
        private const val MAX_THINGS = 64
        private const val TAG = "GL Engine"
    }
}
